Enables SharedVariableAtomicityDetector

kevinrr888 opened the pull request:
Enables the SharedVariableAtomicityDetector SpotBugs detector. Resolved all race conditions and all false-positives (with suppressions or NotThreadSafe annotation)

closes #5636

kevinrr888 committed:
Enables SharedVariableAtomicityDetector

Enables the SharedVariableAtomicityDetector SpotBugs detector.
Resolved all race conditions and all false-positives (with suppressions or
NotThreadSafe annotation)

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchReaderIterator.java:
```diff
@@ -355,6 +357,7 @@ private String getTableInfo() {
   }
 
+  @NotThreadSafe
   private class QueryTask implements Runnable {
```
QueryTask is not used as/expected to be a thread-safe object

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/file/map/MapFileOperations.java:
```diff
@@ -43,6 +45,7 @@
   private static final String MSG = "Map files are not supported";
 
+  @NotThreadSafe
   public static class RangeIterator implements FileSKVIterator {
```
accumulo iterators are not expected to be thread safe

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/SimpleBufferedOutputStream.java:
```diff
@@ -22,10 +22,13 @@
  * been buffered.
  */
+@NotThreadSafe
 class SimpleBufferedOutputStream extends FilterOutputStream {
```
The fact that this is a stream and the the impl itself show this is not a thread safe class

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:
```diff
@@ -754,6 +757,7 @@ public long getLength() {
   }
 
+  @NotThreadSafe
   private static class LocalityGroupReader extends LocalityGroup implements FileSKVIterator {
```
accumulo iterators are not expected to be thread safe

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/file/BloomFilterLayer.java:
```diff
@@ -346,6 +348,7 @@ public void close() {
   }
 
+  @NotThreadSafe
   public static class Reader implements FileSKVIterator {
```
accumulo iterators are not expected to be thread safe

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/iteratorsImpl/system/ColumnFamilySkippingIterator.java:
```diff
@@ -36,6 +38,7 @@
 
+@NotThreadSafe
 public class ColumnFamilySkippingIterator extends ServerSkippingIterator
     implements InterruptibleIterator {
```
accumulo iterators are not expected to be thread safe

kevinrr888 commented on server/base/src/main/java/org/apache/accumulo/server/fs/FileManager.java:
```diff
@@ -386,6 +388,7 @@ private void releaseReaders(KeyExtent tablet, List<FileSKVIterator> readers,
   }
 
+  @NotThreadSafe
   static class FileDataSource implements DataSource {
```
A supporting class for SourceSwitchingIterator which is an iterator which is not thread safe, and impl shows this is intended for single threaded access (e.g., no volatile, sync, etc).

kevinrr888 commented on server/base/src/main/java/org/apache/accumulo/server/problems/ProblemReportingIterator.java:
```diff
@@ -33,6 +35,7 @@
 import org.apache.accumulo.server.ServerContext;
 
+@NotThreadSafe
 public class ProblemReportingIterator implements InterruptibleIterator {
```
accumulo iterators are not thread safe

kevinrr888 commented on server/tserver/src/main/java/org/apache/accumulo/tserver/InMemoryMap.java:
```diff
@@ -531,6 +533,7 @@ public synchronized long getNumEntries() {
   private final Set<MemoryIterator> activeIters = Collections.synchronizedSet(new HashSet<>());
 
+  @NotThreadSafe
   class MemoryDataSource implements DataSource {
```
A supporting class for iterators which are not thread safe, and impl shows this is intended for single threaded access (e.g., no volatile, sync, etc).

kevinrr888 commented on server/tserver/src/main/java/org/apache/accumulo/tserver/NativeMap.java:
```diff
@@ -534,6 +536,7 @@ public void delete() {
   }
 
+  @NotThreadSafe
   private static class NMSKVIter implements InterruptibleIterator {
```
Iterators are not thread safe

kevinrr888 reviewed (commented):
Provided my reasoning in case it's helpful in review

kevinrr888 committed:
Merge branch '2.1' into 2.1-feature-5636

kevinrr888 committed:
NotThreadSafe to CountingInputStream

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/DistributedReadWriteLock.java:
```diff
@@ -111,7 +112,7 @@ static class ReadLock implements Lock {
     QueueLock qlock;
     byte[] userData;
-    long entry = -1;
+    AtomicLong entry;
```
Changes may not be needed to this class. I think it is only used by a single thread, but not 100% sure.  Did some sampling and found this usually created as a local var in a function to try to obtain a lock in zookeeper, but the local var is not shared among threads.

keith-turner reviewed (commented):
How does this static analysis pick which classes to inspect?  Does every class that has non volatile and mutable fields have to be marked as not thread safe?

kevinrr888 commented:
> How does this static analysis pick which classes to inspect? Does every class that has non volatile and mutable fields have to be marked as not thread safe?

I believe it looks at every class, but not entirely sure how it chooses what to flag. I think it looks for common non-thread safe patterns and flags those (e.g., getting and then setting an atomic var in two operations, modifying a non-volatile/non-atomic var in a run(), but also having a getter in that same class, etc.). The documentation is pretty vague. https://spotbugs.readthedocs.io/en/latest/detectors.html

It does not seem that every class needs to be marked/considered, I only had to change ~40 for 2.1

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/DistributedReadWriteLock.java:
```diff
@@ -154,22 +155,27 @@ public void lockInterruptibly() throws InterruptedException {
-      SortedMap<Long,byte[]> entries = qlock.getEarlierEntries(entry);
+      var entryVal = entry.updateAndGet(val -> {
+        if (val == -1) {
+          var newVal = qlock.addEntry(new ParsedLock(this.lockType(), this.userData).getLockData());
```
Looking at this again, I agree with you that no changes are needed for this class. When I was reviewing this the first time, I had incorrectly followed a code trace. This class is not intended to be thread safe and threads never share the same ReadLock. So, this can be safely marked not thread safe and I'll revert my changes :+1:

kevinrr888 committed:
revert changes to DistributedReadWriteLock.ReadLock, mark not thread safe instead

kevinrr888 committed:
formatting

kevinrr888 committed:
changes:

* changed `batchTimeout = scanner.getTimeout` to `batchTimeout =
  scanner.getBatchTimeout` in ClientSideIteratorScanner, appeared to be
calling incorrect method
* address review

kevinrr888 committed:
remove NotThreadSafe from Scanner, add suppression

dlmarion reviewed (commented):
Still reviewing....

kevinrr888 committed:
revert a change to ClientSideIteratorScanner in favor of opening a new PR

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/DistributedReadWriteLock.java:
```diff
@@ -107,6 +109,7 @@ public interface QueueLock {
 
   private static final Logger log = LoggerFactory.getLogger(DistributedReadWriteLock.class);
 
+  @NotThreadSafe
```
There has been some discussion around this class: 
https://github.com/apache/accumulo/pull/6025#discussion_r2624917759
https://github.com/apache/accumulo/pull/6025#discussion_r2624926975
Threads never share the same ~~DistributedReadWriteLock~~ ReadLock, so needs to be marked not thread safe as it has concurrency issues if this were to be shared b/w threads

kevinrr888 replied:
Added comment in 0b96b0898ca296e6a2771416854a1e44d7712786

keith-turner commented on server/base/src/main/java/org/apache/accumulo/server/tablets/TabletTime.java:
```diff
@@ -87,6 +89,8 @@ public MetadataTime getMetadataTime(long time) {
     @Override
+    @SuppressFBWarnings(value = "AT_NONATOMIC_64BIT_PRIMITIVE",
+        justification = "this is only called in tablet constructor, so does not need to be done atomically")
     public void useMaxTimeFromWALog(long time) {
```
Would making this method syncronized remove the warning?  Seems safe to add the sync and it makes the code more tolerant of changes in how its used.

```suggestion
    public synchronized void useMaxTimeFromWALog(long time) {
```

Also seems like the following methods in this class should be made synchronized because they read and/or write the classes variables.

- `MetadataTime getMetadataTime()`
- `public long getTime()`

kevinrr888 committed:
Merge branch '2.1' into 2.1-feature-5636

kevinrr888 committed:
address reviews

kevinrr888 committed:
revert CountingIterator back to NotThreadSafe

DomGarguilo commented:
Looked through everything and fixed the things I saw in 24a9abd. Here is a summary of the changes there:

 - GCRun: counters were being incremented in a thread pool so switched to LongAdder for thread safety
 - ‎ThriftTransportPool: CachedTTransport is accessed by two threads by design - suppress the specific scope instead
 - Scanner: fixed a bug where it was possible that a lock was released even if it was not aquired. Added a gate to only unlock if the lock was obtained
 - DefaultCompactionPlanner: the suppression stated that feilds were set once and read only after but init() runs on a different thread so made things volatile for safety
 - added the `<optional>` feild to the `jsr305` declarations in the poms since its never needed at runtime
